{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16896b6d-5ca7-49bf-96b7-68bfc8a42509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading libraries\n",
    "library(Gviz)\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "library(GenomicAlignments)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(BiocManager)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(Biostrings)\n",
    "library(KEGGREST)\n",
    "library(AnnotationDbi)\n",
    "library(GenomicAlignments)\n",
    "library(rtracklayer)\n",
    "library(biomaRt)\n",
    "library(GenomicFeatures)\n",
    "library(BSgenome)\n",
    "library(plotgardener)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c552bc4f-5ae6-446e-bb1f-695f19283e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the region of interest \n",
    "chrom <- \"chr7\"\n",
    "start <- 18086949\n",
    "end <- 19002416"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df40b55-4b81-4304-8d99-2892cb230890",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load GTF file\n",
    "gtf_file <- \"gencode.v44.basic.annotation.gtf\"\n",
    "gtf_data <- import(gtf_file)\n",
    "\n",
    "# Filter gtf_data for the specific chromosome and region\n",
    "subset_gtf_data <- gtf_data[seqnames(gtf_data) == chrom & start(gtf_data) >= start & end(gtf_data) <= end, ]\n",
    "\n",
    "# Additionally filter for the gene of interest\n",
    "gene_of_interest <- \"HDAC9\"\n",
    "subset_gtf_data <- subset_gtf_data[subset_gtf_data$gene_name == gene_of_interest, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99290570-eecf-4e0a-98fa-d06344cbac64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ideogram track ----- \n",
    "gen <- \"hg38\"\n",
    "itrack <- IdeogramTrack(genome = gen, chromosome = chrom, name = '')\n",
    "\n",
    "displayPars(itrack) <- list(showBandId = FALSE, # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                col.frame = 'black', col = 'black', cex = 1.2, col.line = 'black',showTitle = TRUE,\n",
    "                               fontcolor.group = 'black')  # Set the font color to black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02363039-7be9-4d24-9a17-1e41faa901da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For the transcript id plot\n",
    "filtered_gtf_data <- subset_gtf_data[subset_gtf_data$type == \"transcript\"]\n",
    "head(filtered_gtf_data)\n",
    "table(filtered_gtf_data$transcript_id)\n",
    "mcols(filtered_gtf_data)$feature <- filtered_gtf_data$transcript_id \n",
    "gene_track <- GeneRegionTrack(filtered_gtf_data,\n",
    "                              name = \"\",\n",
    "                              chromosome = chrom,\n",
    "                              start = start, end = end,\n",
    "                              transcriptAnnotation = 'feature',\n",
    "                              showId = TRUE, just.group = \"above\",\n",
    "                              fontcolour = 'black',\n",
    "                              fontsize = 300\n",
    "                              ) \n",
    "\n",
    "displayPars(gene_track) <- list(fontcolor = 'black', col = 'black', cex = 10, col.line = 'black',\n",
    "                               fontcolor.group = 'black')  # Set the font color to black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b4cf426-281c-40b7-93d7-02ad6d739aca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For adding the exon information, let's filter the exonic information \n",
    "subset_gtf_transcripts_exons <- subset_gtf_data[subset_gtf_data$type %in% c(\"exon\"), ]\n",
    "head(subset_gtf_transcripts_exons)\n",
    "subset_gtf_transcripts_exons <- subset_gtf_transcripts_exons[subset_gtf_transcripts_exons$transcript_id %in% filtered_gtf_data$transcript_id,]\n",
    "table(subset_gtf_transcripts_exons$transcript_id)\n",
    "\n",
    "exon_track <- AnnotationTrack(subset_gtf_transcripts_exons,\n",
    "                              name = \"Exons\",\n",
    "                              chromosome = chrom,\n",
    "                              start = start, end = end ,\n",
    "                              fill = \"darkgreen\",    # Color for exons\n",
    "                              col = \"black\",\n",
    "                              direction = NA)         # Border color for exons\n",
    "# Ideogram track ----- \n",
    "gen <- \"hg38\"\n",
    "itrack <- IdeogramTrack(genome = gen, chromosome = chrom, name = '')\n",
    "\n",
    "displayPars(itrack) <- list(showBandId = FALSE, # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                col.frame = 'black', col = 'black', cex = 1.2, col.line = 'black',showTitle = TRUE,\n",
    "                               fontcolor.group = 'black')  # Set the font color to black\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf937428-58e5-4816-99f1-b6bc1bdfc73d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For bam coverage plot: Define the bam files for tumor and normal samples\n",
    "\n",
    "# Define the path to the BAM folder and the sample information file\n",
    "bam_folder <- \".\"\n",
    "sample_info_file <- as.data.frame(read_csv('meta_with_batchinfo.csv'))\n",
    "\n",
    "# List all BAM files in the folder\n",
    "bam_files <- list.files(path = bam_folder, pattern = \"*.bam.bai\", full.names = FALSE)\n",
    "\n",
    "# Extract sample IDs from BAM filenames (the \"SRRxxxxxx\" part)\n",
    "bam_sample_ids <- sapply(strsplit(bam_files, \"_\"), `[`, 1)\n",
    "\n",
    "# Add sample IDs to a data frame with BAM file names\n",
    "bam_df <- data.frame(sample_id = bam_sample_ids, bam_file = bam_files, stringsAsFactors = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d78547-12ce-43d4-8441-01101712e9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(sample_info_file)\n",
    "\n",
    "sample_info_file <- sample_info_file %>%\n",
    "  mutate(Tumor_Normal1 = case_when(\n",
    "    Tumor_Normal %in% c(\"TumorTissue\", \"TumorTissue.2018\", \"TumorTissue.rep\") ~ \"Tumor\",\n",
    "    Tumor_Normal %in% c(\"PairedNormalTissue\", \"PairedNormalTissue.rep\") ~ \"AdjNormal\"\n",
    "  ))\n",
    "sample_info_file <- sample_info_file[,c('Run', 'Tumor_Normal1')]\n",
    "\n",
    "# Merge the BAM data frame with the sample information\n",
    "merged_data <- merge(bam_df, sample_info_file, by.x = \"sample_id\", by.y = \"Run\")\n",
    "\n",
    "# Filter for tumor and normal samples\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "\n",
    "tumor_bam_files <- merged_data %>%\n",
    "  filter(Tumor_Normal1 == \"Tumor\") %>%\n",
    "  dplyr::select(bam_file)\n",
    "\n",
    "normal_bam_files <- merged_data %>%\n",
    "  filter(Tumor_Normal1 == \"AdjNormal\") %>%\n",
    "  dplyr::select(bam_file)\n",
    "\n",
    "dim(tumor_bam_files)\n",
    "dim(normal_bam_files)\n",
    "\n",
    "head(tumor_bam_files)\n",
    "head(normal_bam_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1428fc9-cc76-4f94-86cf-a938904627e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the bam files\n",
    "tnbc_bam <- tumor_bam_files$bam_file\n",
    "nor_bam <- normal_bam_files$bam_file\n",
    "\n",
    "head(tnbc_bam)\n",
    "head(nor_bam)\n",
    "\n",
    "# Remove the \".bai\" extension from each filename\n",
    "tnbc_bam <- gsub(\".bai$\", \"\", tnbc_bam)\n",
    "nor_bam <- gsub(\".bai$\", \"\", nor_bam)\n",
    "\n",
    "head(tnbc_bam)\n",
    "head(nor_bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2453bd38-1bcd-4832-83f0-aba60a11af72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modify to compute mean coverage\n",
    "get_mean_coverage_as_granges <- function(bam_files, chromosome, start, end) {\n",
    "  coverage_list <- lapply(bam_files, function(bam_file) {\n",
    "    bam_data <- readGAlignments(bam_file, param=ScanBamParam(which=GRanges(chromosome, IRanges(start, end))))\n",
    "    coverage(bam_data)[[chromosome]]\n",
    "  })\n",
    "  # Sum coverages and then divide by the number of BAM files to get the mean\n",
    "  cumulative_coverage <- Reduce(\"+\", coverage_list)\n",
    "  mean_coverage <- cumulative_coverage / length(bam_files)\n",
    "  \n",
    "  gr <- GRanges(seqnames = chromosome,\n",
    "                ranges = IRanges(start = start:(end - 1), width = 1),\n",
    "                score = as.numeric(mean_coverage[start:(end - 1)]))\n",
    "  return(gr)\n",
    "}\n",
    "\n",
    "# Use the new function to get mean coverage\n",
    "tnbc_coverage_mean <- get_mean_coverage_as_granges(tnbc_bam, chromosome = chrom, start = start, end = end)\n",
    "nor_coverage_mean <- get_mean_coverage_as_granges(nor_bam, chromosome = chrom, start = start, end = end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47fb6221-5b26-4260-ab7f-bf73f1f75aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_values <- mcols(tnbc_coverage_mean)$score\n",
    "quantile(coverage_values, probs = c(0.05,0.5, 0.95))\n",
    "max(coverage_values)\n",
    "coverage_values <- mcols(nor_coverage_mean)$score\n",
    "quantile(coverage_values, probs = c(0.05, 0.5,0.95))\n",
    "max(coverage_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c0c31cc-344a-4451-9ba8-671cccf8cab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TNBC track with explicit color settings for histogram\n",
    "tnbc_track <- DataTrack(range = tnbc_coverage_mean, genome = \"hg38\", type = \"hist\",\n",
    "                         chromosome = chrom, start = (start), end = end,\n",
    "                         name = \"TNBC\\nmean\\ncoverage\", col.histogram = 'tomato2',\n",
    "                         fill.histogram='tomato2', ylim = c(0,40))\n",
    "\n",
    "# Normal track with explicit color settings for histogram\n",
    "nor_track <- DataTrack(range = nor_coverage_mean, genome = \"hg38\", type = \"hist\",\n",
    "                          chromosome = chrom, start = (start), end = end,\n",
    "                          name = \"Normal\\nmean\\ncoverage\", col.histogram = 'steelblue1',\n",
    "                          fill.histogram='steelblue1', ylim = c(0,40))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e86d050-3687-424b-a78d-f0ca8f67ab52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create GenomeAxisTrack\n",
    "axis_track <- GenomeAxisTrack(name = 'Axis Track', add53 = TRUE, add35 = TRUE)\n",
    "displayPars(axis_track) <- list(  # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                fontcolor = 'black', col = 'black', cex = 1, col.line = 'black',\n",
    "                               fontcolor.group = 'black')  # Set the font color to black\n",
    "\n",
    "# Create promoter regions (p1; p2; p3)\n",
    "p1 <- AnnotationTrack(start = 18086949, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"pink\", col = NULL, \n",
    "                     fontcolor.group = 'black',fontcolor.item = 'black')\n",
    "p2 <- AnnotationTrack(start = 18495745, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"orange\", col = NULL,\n",
    "                      fontcolor.group = 'black', fontcolor.item = 'black')\n",
    "p3 <- AnnotationTrack(start = 18509273, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"lightblue\", col = NULL,\n",
    "                      fontcolor.group = 'black', fontcolor.item = 'black')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6682fbfb-5116-44f0-ba11-198a49712ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For adding the H3K4me3 track for Tumor\n",
    "bw.file_tum <- 'HCC1937_t2_SRX1998157.bw'\n",
    "bw_tumor<- readBigwig(file=bw.file_tum,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)\n",
    "head(bw_tumor)\n",
    "summary(bw_tumor$score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8761ce90-d903-4ea3-ad2e-c933d4f84e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tum <- bw_tumor\n",
    "gr_data_tumor <- GRanges(\n",
    "    seqnames = data_tum$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_tum$start,        # Start positions\n",
    "        end = data_tum$end             # End positions\n",
    "    ),\n",
    "    score = data_tum$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_tumor <- DataTrack(\n",
    "    range = gr_data_tumor,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"HCC1937\\nH3K4me3\", # Name of the track\n",
    "    type = \"h\",\n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),# Set limits for the y-axis\n",
    "    col = \"tomato2\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dc13afd-a137-444f-b39d-42874fb54aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For adding the H3K4me3 track for Normal\n",
    "bw.file_nor <- 'MCF10A_t1.bw'\n",
    "bw_normal<- readBigwig(file=bw.file_nor,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)\n",
    "head(bw_normal)\n",
    "summary(bw_normal$score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3aab3c8-844b-4896-8a97-d243172e3039",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_normal <- bw_normal\n",
    "\n",
    "gr_data_normal <- GRanges(\n",
    "    seqnames = data_normal$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_normal$start,        # Start positions\n",
    "        end = data_normal$end             # End positions\n",
    "    ),\n",
    "    score = data_normal$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_nor <- DataTrack(\n",
    "    range = gr_data_normal,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"MCF10A\\nH3K4me3\",\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),       # Set limits for the y-axis\n",
    "    col = \"steelblue1\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95524d6c-661f-43e1-af32-694121e30e21",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Chunck for adding the H3K27ac enhancer chipseq data plot:------------------------->>\n",
    "\n",
    "# For Tumor\n",
    "bw.file_tum <- 'HCC1937.H3K27ac_SRX1998148.bw'\n",
    "bw_tumor<- readBigwig(file=bw.file_tum,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)\n",
    "head(bw_tumor)\n",
    "summary(bw_tumor$score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddcf677d-f4f1-4a7a-a150-0793ddc57372",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tum <- bw_tumor\n",
    "gr_data_tumor <- GRanges(\n",
    "    seqnames = data_tum$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_tum$start,        # Start positions\n",
    "        end = data_tum$end             # End positions\n",
    "    ),\n",
    "    score = data_tum$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_tumor_enhancer <- DataTrack(\n",
    "    range = gr_data_tumor,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"HCC1937\\nH3K27ac\", # Name of the track\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),     # Set limits for the y-axis\n",
    "    col = \"tomato2\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ade729c0-e83f-42a7-bf88-694b409fa9fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add H3K27ac track for Normal\n",
    "bw.file_nor <- 'MCF10A.H3K27ac_SRX1997949.bw'\n",
    "bw_normal<- readBigwig(file=bw.file_nor,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)\n",
    "head(bw_normal)\n",
    "summary(bw_normal$score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abd41789-c7dd-4fa5-9fe8-f9b55736c4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_normal <- bw_normal\n",
    "\n",
    "gr_data_normal <- GRanges(\n",
    "    seqnames = data_normal$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_normal$start,        # Start positions\n",
    "        end = data_normal$end             # End positions\n",
    "    ),\n",
    "    score = data_normal$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_nor_enhancer <- DataTrack(\n",
    "    range = gr_data_normal,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"MCF10A\\nH3K27ac\",\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),     # Set limits for the y-axis\n",
    "    col = \"steelblue1\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d585329-2e7b-4974-a333-a983ac904ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the highlight track plot for HDAC9\n",
    "ht <- HighlightTrack(trackList = list(p1,p2,p3,tnbc_track,nor_track,track_tumor,track_nor,track_tumor_enhancer,track_nor_enhancer),\n",
    "                    start = c(18082249,18491845,18506023), width = c(8000,8000,8000), chromosome = chrom, \n",
    "                     col = \"orange\", fill = \"orange\", inBackground = FALSE, alpha = 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc8b7d6-bd62-4147-9e16-c9cfde5e9873",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot all tracks\n",
    "pdf('may23_HDAC9_pr1078_added_Final.pdf',width = 8,height = 8)\n",
    "# options(repr.plot.width = 8, repr.plot.height = 8)  # width and height in inches\n",
    "plotTracks(list(axis_track, ht),\n",
    "           from = 18086949, to = 18656949, chromosome = chrom,\n",
    "           sizes = c(4,1,1,1,6,6,6,6,6,6),\n",
    "           background.panel = \"white\",\n",
    "           background.title = \"grey39\", col.axis = \"white\",\n",
    "           extend.left = 0.025,\n",
    "           cex.axis = 0.7, cex.title = 1,showTitle = TRUE)\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
