{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "16896b6d-5ca7-49bf-96b7-68bfc8a42509",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biobase'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:MatrixGenerics':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grid':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "\n",
      "Bioconductor version '3.16' is out-of-date; the current release version '3.21'\n",
      "  is available with R version '4.5'; see https://bioconductor.org/install\n",
      "\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 4.2.3\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicAlignments':\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biostrings':\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:XVector':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from 'package:matrixStats':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggplot2' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'tibble' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'tidyr' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'readr' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'purrr' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'stringr' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'forcats' was built under R version 4.2.3\"\n",
      "Warning message:\n",
      "\"package 'lubridate' was built under R version 4.2.3\"\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mBiostrings\u001b[39m::collapse(), \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m      masks \u001b[34mXVector\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m        masks \u001b[34mmatrixStats\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mGenomicAlignments\u001b[39m::first(), \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m         masks \u001b[34mGenomicAlignments\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mGenomicAlignments\u001b[39m::second(), \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m       masks \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mXVector\u001b[39m::slice(), \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: 'plotgardener'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    c\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Loading libraries\n",
    "library(Gviz)\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "library(GenomicAlignments)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(BiocManager)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(Biostrings)\n",
    "library(KEGGREST)\n",
    "library(AnnotationDbi)\n",
    "library(biomaRt)\n",
    "library(GenomicFeatures)\n",
    "library(BSgenome)\n",
    "library(plotgardener)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c552bc4f-5ae6-446e-bb1f-695f19283e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the region of interest \n",
    "chrom <- \"chr7\"\n",
    "start <- 18086949\n",
    "end <- 19002416"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4df40b55-4b81-4304-8d99-2892cb230890",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load GTF file\n",
    "gtf_file <- \"gencode.v44.basic.annotation.gtf\"\n",
    "gtf_data <- import(gtf_file)\n",
    "\n",
    "# Filter gtf_data for the specific chromosome and region\n",
    "subset_gtf_data <- gtf_data[seqnames(gtf_data) == chrom & start(gtf_data) >= start & end(gtf_data) <= end, ]\n",
    "\n",
    "# Additionally filter for the gene of interest\n",
    "gene_of_interest <- \"HDAC9\"\n",
    "subset_gtf_data <- subset_gtf_data[subset_gtf_data$gene_name == gene_of_interest, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "99290570-eecf-4e0a-98fa-d06344cbac64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ideogram track ----- \n",
    "gen <- \"hg38\"\n",
    "itrack <- IdeogramTrack(genome = gen, chromosome = chrom, name = '')\n",
    "\n",
    "displayPars(itrack) <- list(showBandId = FALSE, # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                col.frame = 'black', col = 'black', cex = 1.2, col.line = 'black',showTitle = TRUE,\n",
    "                               fontcolor.group = 'black')  # Set the font color to black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "02363039-7be9-4d24-9a17-1e41faa901da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For the transcript id plot\n",
    "filtered_gtf_data <- subset_gtf_data[subset_gtf_data$type == \"transcript\"]\n",
    "mcols(filtered_gtf_data)$feature <- filtered_gtf_data$transcript_id\n",
    "gene_track <- GeneRegionTrack(filtered_gtf_data,\n",
    "                              name = \"\",\n",
    "                              chromosome = chrom,\n",
    "                              start = start, end = end,\n",
    "                              transcriptAnnotation = 'feature',\n",
    "                              showId = TRUE, just.group = \"above\",\n",
    "                              fontcolour = 'black',\n",
    "                              fontsize = 300\n",
    "                              ) \n",
    "\n",
    "displayPars(gene_track) <- list(fontcolor = 'black', col = 'black', cex = 10, col.line = 'black',\n",
    "                               fontcolor.group = 'black')  # Set the font color to black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b4cf426-281c-40b7-93d7-02ad6d739aca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For adding the exon information, let's filter the exonic information \n",
    "subset_gtf_transcripts_exons <- subset_gtf_data[subset_gtf_data$type %in% c(\"exon\"), ]\n",
    "subset_gtf_transcripts_exons <- subset_gtf_transcripts_exons[subset_gtf_transcripts_exons$transcript_id %in% filtered_gtf_data$transcript_id,]\n",
    "exon_track <- AnnotationTrack(subset_gtf_transcripts_exons,\n",
    "                              name = \"Exons\",\n",
    "                              chromosome = chrom,\n",
    "                              start = start, end = end ,\n",
    "                              fill = \"darkgreen\",    # Color for exons\n",
    "                              col = \"black\",\n",
    "                              direction = NA)         # Border color for exons\n",
    "# Ideogram track ----- \n",
    "gen <- \"hg38\"\n",
    "itrack <- IdeogramTrack(genome = gen, chromosome = chrom, name = '')\n",
    "displayPars(itrack) <- list(showBandId = FALSE, # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                col.frame = 'black', col = 'black', cex = 1.2, col.line = 'black',showTitle = TRUE,\n",
    "                               fontcolor.group = 'black')  # Set the font color to black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf937428-58e5-4816-99f1-b6bc1bdfc73d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m448\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m99\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (45): Project_ID, Run, batch, download_path, Experiment, LibraryName, L...\n",
      "\u001b[32mdbl\u001b[39m  (31): ...1, spots, bases, spots_with_mates, avgLength, size_MB, InsertS...\n",
      "\u001b[33mlgl\u001b[39m  (21): AssemblyName, Study_Pubmed_id, g1k_pop_code, source, g1k_analysis...\n",
      "\u001b[34mdttm\u001b[39m  (2): ReleaseDate, LoadDate\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "# For bam coverage plot: Define the bam files for tumor and normal samples\n",
    "# Define the path to the BAM folder and the sample information file\n",
    "bam_folder <- \".\"\n",
    "sample_info_file <- as.data.frame(read_csv('meta_with_batchinfo.csv'))\n",
    "\n",
    "# List all BAM files in the folder\n",
    "bam_files <- list.files(path = bam_folder, pattern = \"*.bam.bai\", full.names = FALSE)\n",
    "\n",
    "# Extract sample IDs from BAM filenames (the \"SRRxxxxxx\" part)\n",
    "bam_sample_ids <- sapply(strsplit(bam_files, \"_\"), `[`, 1)\n",
    "\n",
    "# Add sample IDs to a data frame with BAM file names\n",
    "bam_df <- data.frame(sample_id = bam_sample_ids, bam_file = bam_files, stringsAsFactors = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e1d78547-12ce-43d4-8441-01101712e9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for tumor and normal samples\n",
    "sample_info_file <- sample_info_file %>%\n",
    "  mutate(Tumor_Normal1 = case_when(\n",
    "    Tumor_Normal %in% c(\"TumorTissue\", \"TumorTissue.2018\", \"TumorTissue.rep\") ~ \"Tumor\",\n",
    "    Tumor_Normal %in% c(\"PairedNormalTissue\", \"PairedNormalTissue.rep\") ~ \"AdjNormal\"\n",
    "  ))\n",
    "sample_info_file <- sample_info_file[,c('Run', 'Tumor_Normal1')]\n",
    "\n",
    "# Merge the BAM data frame with the sample information\n",
    "merged_data <- merge(bam_df, sample_info_file, by.x = \"sample_id\", by.y = \"Run\")\n",
    "tumor_bam_files <- merged_data %>%\n",
    "  filter(Tumor_Normal1 == \"Tumor\") %>%\n",
    "  dplyr::select(bam_file)\n",
    "\n",
    "normal_bam_files <- merged_data %>%\n",
    "  filter(Tumor_Normal1 == \"AdjNormal\") %>%\n",
    "  dplyr::select(bam_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f1428fc9-cc76-4f94-86cf-a938904627e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the TNBC samples bam files\n",
    "tnbc_bam <- tumor_bam_files$bam_file\n",
    "\n",
    "# Define the normal samples bam files\n",
    "nor_bam <- normal_bam_files$bam_file\n",
    "\n",
    "# Remove the \".bai\" extension from each filename\n",
    "tnbc_bam <- gsub(\".bai$\", \"\", tnbc_bam)\n",
    "nor_bam <- gsub(\".bai$\", \"\", nor_bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2453bd38-1bcd-4832-83f0-aba60a11af72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modify to compute mean coverage\n",
    "get_mean_coverage_as_granges <- function(bam_files, chromosome, start, end) {\n",
    "  coverage_list <- lapply(bam_files, function(bam_file) {\n",
    "    bam_data <- readGAlignments(bam_file, param=ScanBamParam(which=GRanges(chromosome, IRanges(start, end))))\n",
    "    coverage(bam_data)[[chromosome]]\n",
    "  })\n",
    "  # Sum coverages and then divide by the number of BAM files to get the mean\n",
    "  cumulative_coverage <- Reduce(\"+\", coverage_list)\n",
    "  mean_coverage <- cumulative_coverage / length(bam_files)\n",
    "  \n",
    "  gr <- GRanges(seqnames = chromosome,\n",
    "                ranges = IRanges(start = start:(end - 1), width = 1),\n",
    "                score = as.numeric(mean_coverage[start:(end - 1)]))\n",
    "  return(gr)\n",
    "}\n",
    "\n",
    "# Use the new function to get mean coverage\n",
    "tnbc_coverage_mean <- get_mean_coverage_as_granges(tnbc_bam, chromosome = chrom, start = start, end = end)\n",
    "nor_coverage_mean <- get_mean_coverage_as_granges(nor_bam, chromosome = chrom, start = start, end = end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0c0c31cc-344a-4451-9ba8-671cccf8cab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TNBC track with explicit color settings for histogram\n",
    "tnbc_track <- DataTrack(range = tnbc_coverage_mean, genome = \"hg38\", type = \"hist\",\n",
    "                         chromosome = chrom, start = (start), end = end,\n",
    "                         name = \"TNBC\\nmean\\ncoverage\", col.histogram = 'tomato2',\n",
    "                         fill.histogram='tomato2', ylim = c(0,40))\n",
    "\n",
    "# Normal track with explicit color settings for histogram\n",
    "nor_track <- DataTrack(range = nor_coverage_mean, genome = \"hg38\", type = \"hist\",\n",
    "                          chromosome = chrom, start = (start), end = end,\n",
    "                          name = \"Normal\\nmean\\ncoverage\", col.histogram = 'steelblue1',\n",
    "                          fill.histogram='steelblue1', ylim = c(0,40))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e86d050-3687-424b-a78d-f0ca8f67ab52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create GenomeAxisTrack\n",
    "axis_track <- GenomeAxisTrack(name = 'Axis Track', add53 = TRUE, add35 = TRUE)\n",
    "displayPars(axis_track) <- list(  # Set the desired font size (10 is the default; adjust as needed)\n",
    "                                fontcolor = 'black', col = 'black', cex = 1, col.line = 'black',\n",
    "                               fontcolor.group = 'black')  # Set the font color to black\n",
    "\n",
    "# Create promoter regions (p1; p2; p3)\n",
    "p1 <- AnnotationTrack(start = 18086949, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"pink\", col = NULL, \n",
    "                     fontcolor.group = 'black',fontcolor.item = 'black')\n",
    "p2 <- AnnotationTrack(start = 18495745, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"orange\", col = NULL,\n",
    "                      fontcolor.group = 'black', fontcolor.item = 'black')\n",
    "p3 <- AnnotationTrack(start = 18509273, end = end, chromosome = chrom,\n",
    "                      name = \"\", fill = \"lightblue\", col = NULL,\n",
    "                      fontcolor.group = 'black', fontcolor.item = 'black')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6682fbfb-5116-44f0-ba11-198a49712ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For adding the H3K4me3 track for Tumor\n",
    "bw.file_tum <- 'HCC1937_t2_SRX1998157.bw'\n",
    "bw_tumor<- readBigwig(file=bw.file_tum,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8761ce90-d903-4ea3-ad2e-c933d4f84e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tum <- bw_tumor\n",
    "gr_data_tumor <- GRanges(\n",
    "    seqnames = data_tum$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_tum$start,        # Start positions\n",
    "        end = data_tum$end             # End positions\n",
    "    ),\n",
    "    score = data_tum$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_tumor <- DataTrack(\n",
    "    range = gr_data_tumor,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"HCC1937\\nH3K4me3\", # Name of the track\n",
    "    type = \"h\",\n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),# Set limits for the y-axis\n",
    "    col = \"tomato2\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5dc13afd-a137-444f-b39d-42874fb54aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For adding the H3K4me3 track for Normal\n",
    "bw.file_nor <- 'MCF10A_t1.bw'\n",
    "bw_normal<- readBigwig(file=bw.file_nor,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b3aab3c8-844b-4896-8a97-d243172e3039",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_normal <- bw_normal\n",
    "\n",
    "gr_data_normal <- GRanges(\n",
    "    seqnames = data_normal$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_normal$start,        # Start positions\n",
    "        end = data_normal$end             # End positions\n",
    "    ),\n",
    "    score = data_normal$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_nor <- DataTrack(\n",
    "    range = gr_data_normal,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"MCF10A\\nH3K4me3\",\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),       # Set limits for the y-axis\n",
    "    col = \"steelblue1\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "95524d6c-661f-43e1-af32-694121e30e21",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Chunck for adding the H3K27ac enhancer chipseq data plot:------------------------->>\n",
    "\n",
    "# For Tumor\n",
    "bw.file_tum <- 'HCC1937.H3K27ac_SRX1998148.bw'\n",
    "bw_tumor<- readBigwig(file=bw.file_tum,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ddcf677d-f4f1-4a7a-a150-0793ddc57372",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tum <- bw_tumor\n",
    "gr_data_tumor <- GRanges(\n",
    "    seqnames = data_tum$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_tum$start,        # Start positions\n",
    "        end = data_tum$end             # End positions\n",
    "    ),\n",
    "    score = data_tum$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_tumor_enhancer <- DataTrack(\n",
    "    range = gr_data_tumor,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"HCC1937\\nH3K27ac\", # Name of the track\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),     # Set limits for the y-axis\n",
    "    col = \"tomato2\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ade729c0-e83f-42a7-bf88-694b409fa9fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add H3K27ac track for Normal\n",
    "bw.file_nor <- 'MCF10A.H3K27ac_SRX1997949.bw'\n",
    "bw_normal<- readBigwig(file=bw.file_nor,\n",
    "                      chrom = \"chr7\",\n",
    "                      chromstart = 18086949,\n",
    "                      chromend = 19002416)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "abd41789-c7dd-4fa5-9fe8-f9b55736c4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_normal <- bw_normal\n",
    "gr_data_normal <- GRanges(\n",
    "    seqnames = data_normal$seqnames,         # Chromosome column\n",
    "    ranges = IRanges(\n",
    "        start = data_normal$start,        # Start positions\n",
    "        end = data_normal$end             # End positions\n",
    "    ),\n",
    "    score = data_normal$score             # Additional metadata column\n",
    ")\n",
    "\n",
    "## Create a DataTrack object\n",
    "track_nor_enhancer <- DataTrack(\n",
    "    range = gr_data_normal,\n",
    "    genome = \"hg38\",      # Specify the genome version, e.g., \"hg19\" or \"hg38\"\n",
    "    name = \"MCF10A\\nH3K27ac\",\n",
    "    type = \"h\",  \n",
    "    chromosome = chrom,\n",
    "    ylim = c(0, 1.5),     # Set limits for the y-axis\n",
    "    col = \"steelblue1\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8d585329-2e7b-4974-a333-a983ac904ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the highlight track plot for HDAC9\n",
    "ht <- HighlightTrack(trackList = list(p1,p2,p3,tnbc_track,nor_track,track_tumor,track_nor,track_tumor_enhancer,track_nor_enhancer),\n",
    "                    start = c(18082249,18491845,18506023), width = c(8000,8000,8000), chromosome = chrom, \n",
    "                     col = \"orange\", fill = \"orange\", inBackground = FALSE, alpha = 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc8b7d6-bd62-4147-9e16-c9cfde5e9873",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot all tracks\n",
    "pdf('HDAC9_trackplot.pdf',width = 8,height = 8)\n",
    "plotTracks(list(axis_track, ht),\n",
    "           from = 18086949, to = 18656949, chromosome = chrom,\n",
    "           sizes = c(4,1,1,1,6,6,6,6,6,6),\n",
    "           background.panel = \"white\",\n",
    "           background.title = \"grey39\", col.axis = \"white\",\n",
    "           extend.left = 0.025,\n",
    "           cex.axis = 0.7, cex.title = 1,showTitle = TRUE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5548840f-d366-4251-870b-3c860d6f9e4d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
